common/queue.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+// node represents a node in the queue.
+type node[T any] struct {
+	value T
+	next  *node[T]
+}
+
+// Queue is a generic queue implementation using a linked list.
+type Queue[T any] struct {
+	start, end *node[T]
+	size       int
+}
+
+// NewQueue creates a new empty queue.
+func NewQueue[T any]() *Queue[T] {
```

Comment:
There are two ways to instantiate a Queue object currently:
* `common.Queue{}`
* `common.NewQueue()`

Both do the same thing currently, but it's something that can cause confusion as in what's the right way. And, it's also difficult to ensure that both the approaches are always semantically equivalent in the future. So, either remove the NewQueue method altogether or ensure that Queue cannot be initialized by another means except NewQueue as in:

```
type Queue[T any] interface {
	Enqueue(a T)
   ... other methods ...
}

type linkedListBasedQueue[T any] struct {
  ...
}

func (l *linkedListBasedQueue[T]) Enqueue(a T) {
	fmt.Println("Hello")
}

func NewQueue[T any]() Queue[T] {
	return &linkedListBasedQueue[T]{}
}

```